helix-term/src/ui/mod.rs

```diff
@@ -265,6 +266,76 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> FilePi
     picker
 }
 
+pub fn file_browser(root: PathBuf) -> Result<FilePicker, std::io::Error> {
+    let root = helix_stdx::path::canonicalize(root);
+    let directory_content = directory_content(&root)?;
+
+    let columns = [PickerColumn::new(
+        "path",
+        |item: &PathBuf, root: &PathBuf| {
+            let name = item.strip_prefix(root).unwrap_or(item).to_string_lossy();
+            if item.is_dir() {
+                format!("{}/", name).into()
+            } else {
+                name.into()
+            }
+        },
+    )];
+    let picker = Picker::new(
+        columns,
+        0,
+        directory_content,
+        root,
+        move |cx, path: &PathBuf, action| {
+            if path.is_dir() {
+                let owned_path = path.clone();
+                let callback = Box::pin(async move {
+                    let call: Callback =
+                        Callback::EditorCompositor(Box::new(move |_editor, compositor| {
+                            if let Ok(picker) = file_browser(owned_path) {
+                                compositor.push(Box::new(overlay::overlaid(picker)));
+                            }
+                        }));
+                    Ok(call)
+                });
+                cx.jobs.callback(callback);
+            } else if let Err(e) = cx.editor.open(path, action) {
+                let err = if let Some(err) = e.source() {
+                    format!("{}", err)
+                } else {
+                    format!("unable to open \"{}\"", path.display())
+                };
+                cx.editor.set_error(err);
+            }
+        },
+    )
+    .with_preview(|_editor, path| Some((path.as_path().into(), None)));
+
+    Ok(picker)
+}
+
+fn directory_content(path: &Path) -> Result<Vec<PathBuf>, std::io::Error> {
+    let mut dirs = Vec::new();
+    let mut files = Vec::new();
+    for entry in std::fs::read_dir(path)?.flatten() {
+        if entry.path().is_dir() {
```

Comment:
We should try to avoid `Path::is_dir` (and equivalents like `PathBuf::is_dir`) since they mean a syscall per call. We can avoid that here (most of the time, platform dependent) by using `DirEntry::file_type`:

```suggestion
        if entry.file_type().is_ok_and(|file_type| file_type.is_dir()) {
```

I would also switch the return Vec from this function to be `(PathBuf, bool)`, so we keep that `is_dir` boolean around rather than calling `path.is_dir()` elsewhere (especially while rendering). This will make the file browser type different than FilePicker:

```rust
type FileBrowser = Picker<(PathBuf, bool), PathBuf>;
```